### ISSUE TITLE
FPS / video bit rate quickly dropping to 0 while streaming on cellular LTE

### ISSUE THREAD

**Issue by jhays** [shogo4405/HaishinKit.swift#1203]
### Describe the bug

I am attempting to stream video over an LTE network, using RTMP and haishinKit. 
When I run a speed test from the phone with cellular connectivity, I am seeing 1.9Mbps down / 4.9 Mbps up. 

For my current test I am only trying to stream video with 240x486 resolution, 30fps, 500Kbps video bit rate. 

Given the speed I have, and the low quality video I am trying to send, shouldn't this be possible? 

But I am frequently seeing the stream start, and then the FPS and bit rate is quickly plummeting both to 0, and then not ever recovering. 

I'm struggling to understand why, or if there are any settings I can change to increase the stability, and be able to broadcast live video for longer periods of time. 

Ultimately I would like to increase the resolution to 720 and increase the bitrate, but I'm starting my tests with very low quality as I try to figure out where the stability issue is.

### To Reproduce

1. Start broadcasting via RTMP while on a cellular network
2. Observe stream starts successfully
3. Observe logging of FPS / currentBytesInPerSecond / currentBytesOutPerSecond
4. FPS and bytes out will quickly both plummet to 0, and do not recover
5. Sometimes observe a connectClosed event. Not always- sometimes it continues streaming but with 0 fps and 0 bitrate

### Expected behavior

I believe video streaming should be possible with the conditions, and fps / bit rate should not plummet and stay at 0.

### Version

1.4.2

### Smartphone info.

iPhone 14 Pro Max
iOS 16.4.1 

### Additional context

My logging pasted below includes both the haishinKit trace logging, and I have an NSLog line firing ever 1 second that prints out the `rtmpStream.currentFPS`, and `rtmpConnection.currentBytesInPerSecond` and `rtmpConection.currentBytesOutPerSecond`

If you search for that line in the log (the lines start with "currentFPS:" and fire every second), you will see the FPS starts at 30, then eventually drops to 0 and stays stuck. Video bitrate then also drops to 0 and the stream does not recover. 

### Screenshots

This is a graph from the Mux dashboard, where I am streaming the video up to. You can observe their graph also sees the video bit rate and fps dropping to 0 and not recovering. 

<img width="377" alt="haishinKit_LTE_mux_graph" src="https://github.com/shogo4405/HaishinKit.swift/assets/125045/ae13eefb-63b6-42a1-bb6e-172fb0005f71">

### Relevant log output

```shell
Log was too long to post. Please see link to log: 
https://drive.google.com/file/d/1ae9BejtuzWG_3XMBcFso8EUR8p98gDRL/view?usp=share_link
```

**Comment by shogo4405** [comment]
I have reviewed the logs and identified the following issues:

1. The LTE connection is being disconnected a few seconds after starting the stream. This appears to be a problem related to the mobile carrier or the device itself.
1. The disconnection is not being detected even when it occurs.
Regarding the first issue, it seems challenging to resolve it directly within the framework as it is likely influenced by limitations imposed by the mobile carrier or the device.

Regarding the second issue, one possible approach is to implement a mechanism that detects abnormal connection status when there is no response from the server for a certain period of time and handles it by closing the connection.

**Comment by shogo4405** [comment]
Which mobile carrier are you referring to? In the past, there have been similar issues reported with Vodafone in Europe. In those cases, the problem was resolved by using rtmps with SSL. In Japan, some carriers have implemented controls on RTMP to maintain lower prices.

**Comment by jhays** [comment]
Thank you very much for the quick response and looking over my log. 

In my test, I was using Verizon as the carrier, in the US. 
I think I did test rtmps once after I saw that suggestion in another GitHub issue thread, but in my case it didn't not seem to resolve the issue. 

I have tested on the same phone & carrier successfully when stronger signal is available, and I have performed successful streams where the issue hasn't occurred. So I don't think the carrier is blocking this traffic.

I'm thinking I may want to implement automatic reconnection logic so streams can recover, including situations like this.

Would it make sense to poll or monitor `rtmpStream.currentFPS` and `rtmpConection.currentBytesOutPerSecond` at some interval, and if I see they dive to 0, perhaps then I stop the broadcast and attempt to reconnect? Are there any other properties I should be monitoring to determine the health of the broadcast or the bandwidth available? 

When setting up auto-reconnection, I have been setting both rtmpConnection and rtmpStream variables to nil, then instantiating new ones before reconnecting and publishing video to the same server url. Is the correct way to implement reconnection? Once a stream has been closed, is it necessary to reinstantiate rtmpConnection & rtmpStream? I just want to double check that I'm approaching reconnection strategy in the right way. 

Thank you

**Comment by shogo4405** [comment]
Please use this delegate method. https://github.com/shogo4405/HaishinKit.swift/blob/main/Sources/RTMP/RTMPConnection.swift#L30-L33C97

To reconnect, there is no need to set it to nil. Instead, you can close the connection and then initiate a new connection using the connect method. Regarding the interval, it may vary depending on the business requirements, so I will leave the decision up to you.

**Comment by jhays** [comment]
Following up on my previous comment to note that I was able to implement successful reconnection by fully tearing down the rtmpConnection & rtmpStream, closing them, setting the references of each to nil, instantiating new objects for both of those, then connecting and publishing to the same RTMP URL. 

So far I think that this will be a reasonable solution for my use case, but my testing does seem to indicate that there is still an issue with trying to reconnection to an rtmpConnection that was previously closed, or re-publishing on an rtmpStream that was closed. Not sure if the issue lies in one of those or both.

**Comment by shogo4405** [comment]
My recommended retry handling is as shown in the sample program.
https://github.com/shogo4405/HaishinKit.swift/blob/main/Examples/iOS/LiveViewController.swift#L202-L224

**Comment by mkrn** [comment]
Getting similar problems now...
Experiencing 2 issues mainly:

1) These errors on connection failures are not caught, so I can't initiate a reconnect:

2023-05-26 12:24:29.461280+0100 appnew[27463:5354871] [connection] nw_read_request_report [C14] Receive failed with error "Socket is not connected"
2023-05-26 12:24:29.469466+0100 appnew[27463:5354871] [] nw_path_evaluator_create_flow_inner NECP_CLIENT_ACTION_ADD_FLOW 4F47FC11-91DD-41A9-B67E-1943F37E611A [65: No route to host]
2023-05-26 12:24:29.469578+0100 appnew[27463:5354871] [connection] nw_endpoint_flow_setup_channel [C19.1.1 209.38.220.162:443 in_progress channel-flow (satisfied (Path is satisfied), viable, interface: en0[802.11], ipv4, ipv6, dns)] failed to request add nexus flow

2023-05-26 12:24:47.997396+0100 appnew[27463:5353408] [tcp] tcp_output [C14.1.1:2] flags=[R.] seq=2195145339, ack=3515632910, win=2048 state=CLOSED rcv_nxt=3515632910, snd_una=2195058251


2) After re-connection it seems this function is getting very low currentBytesOutPerSecond 

func connection(_ connection: HaishinKit.RTMPConnection, publishSufficientBWOccured stream: HaishinKit.RTMPStream) {

**Comment by mkrn** [comment]
Reproduction of the issues above:
- In Sample app (SwiftUI iOS)
- Start with Wifi + 5G  
- Swipe to controls,  turn off WiFi 
- I'm getting these log messages:

2023-05-26 14:30:25.853018+0100 Example iOS+SwiftUI[27784:5417682] [connection] nw_resolver_start_query_timer_block_invoke [C2] Query fired: did not receive all answers in time for go.eventlive.pro:443
2023-05-26 14:30:42.256915+0100 Example iOS+SwiftUI[27784:5417687] [connection] nw_read_request_report [C1] Receive failed with error "Operation timed out"
2023-05-26 14:30:42.259662+0100 Example iOS+SwiftUI[27784:5417687] [connection] nw_flow_add_write_request [C1 209.38.220.162:443 failed parent-flow (satisfied (Path is satisfied), interface: pdp_ip0[endc_sub6], ipv4, dns, expensive)] cannot accept write requests
2023-05-26 14:30:42.260491+0100 Example iOS+SwiftUI[27784:5417687] [connection] nw_write_request_report [C1] Send failed with error "Socket is not connected"
2023-26-05 14:30:42.262 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
NetConnection.Connect.Closed


17 seconds passes until NetConnection.Connect.Closed is triggered. 

All this time 30FPS is shown 

Can the timeout be regulated or the NetConnection.Connect.Closed  triggered sooner?

**Comment by shogo4405** [comment]
> Can the timeout be regulated or the NetConnection.Connect.Closed triggered sooner?

It has been tuned and will not go any faster. This is talking about NSIntputStream/NSOutputStream conneciton.
But Networkd.framework, It might be a solution. Please set `conneciton.requireNetworkframework = true`.
It works ok.
```
2023-26-05 22:50:27.669 [Info] [com.haishinkit.Exsample.iOS] [LiveViewController.swift:208] rtmpStatusHandler(_:) > NetConnection.Connect.Success
2023-26-05 22:50:28.116 [Info] [com.haishinkit.Exsample.iOS] [LiveViewController.swift:208] rtmpStatusHandler(_:) > NetStream.Publish.Start
2023-26-05 22:50:37.525 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
2023-05-26 22:50:37.525363+0900 Example iOS[20696:5200440] [] nw_endpoint_flow_fillout_data_transfer_snapshot copy_info() returned NULL
2023-05-26 22:50:37.525927+0900 Example iOS[20696:5200440] [connection] nw_connection_copy_connected_local_endpoint_block_invoke [C1] Connection has no local endpoint
2023-26-05 22:50:37.526 [Info] [com.haishinkit.Exsample.iOS] [LiveViewController.swift:208] rtmpStatusHandler(_:) > NetConnection.Connect.Closed
2023-26-05 22:50:38.878 [Info] [com.haishinkit.Exsample.iOS] [LiveViewController.swift:208] rtmpStatusHandler(_:) > NetConnection.Connect.Success
2023-26-05 22:50:39.213 [Info] [com.haishinkit.Exsample.iOS] [LiveViewController.swift:208] rtmpStatusHandler(_:) > NetStream.Publish.Start
```

**Comment by mkrn** [comment]
Thank you yes  it might help..
An issue I have is after reconnection the way publishSufficientBWOccured  is quite different: 
It's getting sufficient but currentBytesOut are consistently lower,
For example,  setting bitrate to 2765000 (2.7 mbps) consistently pushing out ~501216 bits per second (500kbps)

**Comment by shogo4405** [comment]
@mkrn Can you reproduce this way? 500kbps or so even with higher bitrate.
1. In Sample app (SwiftUI iOS)
1. Start with Wifi + 5G
1. Swipe to controls, turn off WiFi
1. I'm getting these log messages:

**Comment by mkrn** [comment]
It reconnected a few times then pushing at very low rate:
```
2023-26-05 15:49:37.124 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
2023-05-26 15:49:37.124498+0100 Example iOS+SwiftUI[27940:5454509] [] nw_endpoint_flow_fillout_data_transfer_snapshot copy_info() returned NULL
2023-05-26 15:49:37.124524+0100 Example iOS+SwiftUI[27940:5454509] [connection] nw_connection_copy_connected_local_endpoint_block_invoke [C1] Connection has no local endpoint
NetConnection.Connect.Closed
NetConnection.Connect.Success
NetStream.Publish.Start
2023-05-26 15:49:40.551039+0100 Example iOS+SwiftUI[27940:5454502] [tcp] tcp_input [C3.1.1:2] flags=[R] seq=1062706553, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=1062706553, snd_una=4216883869
2023-26-05 15:49:40.554 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
NetConnection.Connect.Closed
NetConnection.Connect.Success
NetStream.Publish.Start
2023-05-26 15:49:43.709708+0100 Example iOS+SwiftUI[27940:5454502] [tcp] tcp_input [C4.1.1:2] flags=[R] seq=2243434780, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=2243434780, snd_una=1121857020
2023-26-05 15:49:43.713 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
NetConnection.Connect.Closed
NetConnection.Connect.Success
NetStream.Publish.Start
2023-05-26 15:49:47.358364+0100 Example iOS+SwiftUI[27940:5454505] [tcp] tcp_input [C5.1.1:2] flags=[R] seq=43237643, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=43237643, snd_una=139294224
2023-26-05 15:49:47.362 [Warn] [com.haishinkit.HaishinKit] [RTMPConnection.swift:406] on(status:) > 
NetConnection.Connect.Closed
NetConnection.Connect.Success
NetStream.Publish.Start
NetStream.Publish.Start
2023-05-26 15:50:29.273284+0100 Example iOS+SwiftUI[27940:5454505] [connection] nw_connection_add_timestamp_locked_on_nw_queue [C6] Hit maximum timestamp count, will start dropping events
```

**Comment by shogo4405** [comment]
Since the main concern of this issue is the streaming interruption during the process, let's create a separate issue specifically dedicated to that.

**Comment by jhays** [comment]
Ultimately, for my issue, I am now thinking that it really is just due to low bandwidth on poor LTE connections. 

For now, I'm handling the reconnection process by fully closing & nilling out both rtmpStream and rtmpConnection objects, then creating new ones to perform a new connection to the same rtmp url. Although this does cause a slight visual redraw of the camera preview, it seems to be the most successful path for reconnections for my use case for now. 

Given that I think my initial issue was actually due to bandwidth issues (or maybe event cellular network throttling), I don't think it is an issue with HaishinKit. This issue can be closed if you'd like.

**Comment by mkrn** [comment]
There's definitely a problem with reusing rtmpConnection even after stop publish / start publish it seems to have less bandwidth..  
Nilling both rtmpStream and rtmpConnection wouldn't look as well because of need to re-attach camera @jhays ?

**Comment by jhays** [comment]
@mkrn Yes, in my case I am forced to re-attach the camera with the new rtmpStream and redraw the camera preview. It ends up being a momentary flash of the preview layer, it's not ideal but I can live with it specifically during the reconnection process. Let's hope my project manager & designer can live with it as well...

**Comment by mkrn** [comment]
@shogo4405 @jhays  yes, my reconnects are not working at all now when switching from wifi->lte  or turning lte off and on .. unfortunately when closing rtmpConnection and restarting manually the connection speed is much worse.. (same spot, same network) 
Only when destroying the whole component and  opening again, the connection is back to normal ..

**Comment by shogo4405** [comment]
If it's after reconnection, there is a possibility that the output of VideoToolbox is not stable. This is a characteristic of ABR, and I believe it takes about a minute for the output to stabilize. If this is true, there is a possibility that it might work by discarding the instance of VideoToolbox, so I will try that.

**Comment by levs42** [comment]
@mkrn what device, iOS version and mobile carrier do you use? Trying to reproduce that reconnection bitrate drop. If possible, please add the info to #1210
